### PR TITLE
Add support to whitelist selector expressions

### PR DESCRIPTION
**Pull request by bombsimon**
This will allow a pre defined list of selector expressions X and Sel
idents to mark as whitelisted. By doing this we could allow a specific
type or function from a specific package such as `regexp.MustCompile`.

---

I had a discussion today regarding how to handle things such as `regexp.MustCompile()` or pre defined errors as a global variable which this linter doesn't really support. I know you can ignore errors by using specific variable names but I also saw that there were some `TODO` makers to skip fake errors. I also found #6 which addresses the same issue I have and that you were open to PRs.

I guess I could and should've asked in the ticket but I thought I had a go at an implementation to handle this. Since you're already open to match the name of the variable, would you be open to define a list of selector expression value that could be whitelisted?

As seen in the PR I created a 2D array which holds an `X` ident and a `Sel` ident which will be matched. I added three defaults to test this; `errors New`, `fmt Errorf` and `regexp MustCompile`. If this is a feasible solution this could even be configurable to allow any expressions, like `http Client` I use in the test. Regarding errors specific I guess a check if it implements the `Error()` interface would be better but I still see a use case for a feature like this.

I added a new test file to show the behaviour and also updated the existing tests. Please let me know what you think! If this is not the way you want to go I totally understand and will close this PR.

### CODE REVIEW

File: check_no_globals.go [hunk 1]
```diff
@@ -10,8 +10,48 @@ import (
...
+			break
+		}
+
+		return isWhitelsiteSelectorExpression(expr)
+	}
+
+	return false
+}
+
+func isWhitelsiteSelectorExpression(v *ast.SelectorExpr) bool {
```
Comment (akarl) [inline check_no_globals.go]: Spelling error :)

Comment (bombsimon) [reply check_no_globals.go]: 🤦‍♀️

File: check_no_globals.go [hunk 2]
```diff
@@ -10,8 +10,42 @@ import (
...
+	}
+
+	allowedCombinations := [][]string{
+		{"errors", "New"},
+		{"fmt", "Errorf"},
+		{"regexp", "MustCompile"},
+	}
+
+	for _, i := range allowedCombinations {
+		if x.Name == i[0] && v.Sel.Name == i[1] {
```
Comment (leighmcculloch) [inline check_no_globals.go]: Could we make `i` a type with fields `Name` and `SelName` so the above combinations reads clearer and we about the indexes here?

Comment (bombsimon) [reply check_no_globals.go]: Yeah, sounds like a good idea! I'll fix that!

**Comment by bombsimon** [comment]
I tried to look at your other code to find what style or likings you've got. Feel free to nit on formatting, newlines, naming and whatnot! Also tell me if you wan't me to squash this into a single commit!

**Comment by leonklingele** [comment]
Just follow the [Go convention](https://github.com/golang/go/wiki/Errors#naming) of prefixing global error variables with `"Err"` or `"err"`. IMO there's no need to whitelist error variables using a different naming scheme.

What's the benefit of whitelisting `regexp.MustCompile` given that [the regex won't be checked at runtime](https://github.com/leighmcculloch/gochecknoglobals/issues/6#issuecomment-490286863)?

**Comment by bombsimon** [comment]
Yeah I'm all for following the naming convention for errors and only allow names staring with `Err` or `err`, although it might be a good idea to check both things. This example isn't really an error and shouldn't be an allowed global variable:

```
var ErrorsAllowed = 1
```

Regarding `regexp.MustCompile` it's not so much for the compilation or error checking rather than it's not really efficient to re-compile the regexp each time it's used. Consider calling this function 10000 times per minute:

```
func F(in []byte) [][]byte {
    re := regexp.MustCompile(`^\d+$`)

    return re.FindAll(in, -1)
}
```

versus

```
var re = regexp.MustCompile(`^\d+$`)

func F(in []byte) [][]byte {
    return re.FindAll(in, -1)
}
```

I noticed now that one of the main reasons to add support for this according to the issue was to check for compilation errors but the reason I made this PR was to avoid re-compiling the regexp more than once.

**Comment by leonklingele** [comment]
Why need to call `regexp.MustCompile` multiple times? Just call it once and store its result:

```go
func NewServer() *Server {
	re := regexp.MustCompile(`^\d+$`)
	return &Server{
		re: re,
	}
}

// Then use and inject (*Server).re anywhere you like
```

**Comment by bombsimon** [comment]
How about code with functions requiring regexp that is not attached to a struct? In my case I don't always have a `Server` (or similar). It might be a library with helper functions that don't need a struct.

And even if I did, I don't think I like the idea of adding fields to a struct just to be able to make them "global". The same way I don't like to implement methods on a struct if the struct itself is never used.

But it boils down to a matter of opinion I guess. You could always argue to refactor your code to avoid the need for global variables I guess.

**Comment by leonklingele** [comment]
Attaching the `*regexp.Regexp` to a struct was done to illustrate that there's always a way to somehow store it without having to call `regexp.MustCompile` again.
If you don't have a struct available, simply pass `*regexp.Regexp` just like you pass a `context.Context`.

**Comment by akarl** [comment]
This seems to boil down only to opinions and tastes. Compiled regexps as constants or not.

I personally would like to have compiled regexps as constants. But not everyone agrees.

Could we make the linter configurable. Perhaps providing a list of allowed prefixes or types?

**Comment by bombsimon** [comment]
Yeah I got that it was an example, that’s why I ended my previous post the way I did. I also tried to show other examples which might be harder to refactor for. It wouldn't be much of a helping function if you were to provide the regexp executed yourself.

I’ll leave the PR open and let a maintainer chose if they want to close the PR. 

Thanks for your input!

**Comment by cnmcavoy** [comment]
@leighmcculloch are you going to merge this PR in or did you decide not to support this use case? If you didn't decide to support this behavior, closing this PR would be nice, because its very unclear what state this is in, as an observer 1yr later.

**Comment by leonklingele** [comment]
I still don't like the idea of whitelisting any function calls. `errors.New` and `fmt.Errorf` is already whitelisted, just prefix the variable with `err` or `Err` — stick to the recommended Go coding style. Having a global `*regexp.Regexp` produced by `regexp. MustCompile` is bad style as well. See my previous comments: https://github.com/leighmcculloch/gochecknoglobals/pull/12#issuecomment-490288218, https://github.com/leighmcculloch/gochecknoglobals/pull/12#issuecomment-490538597 & https://github.com/leighmcculloch/gochecknoglobals/pull/12#issuecomment-490549995.

**Comment by cnmcavoy** [comment]
Regular expressions are often used in the same was as constants are in code, so declaring them at the var level in one package, then referencing the var in other packages is a fairly common code pattern. The [documentation for `MustCompile`](https://github.com/golang/go/blob/782fcb44b96753e6e763be91a72bcbfc3dff7ab9/src/regexp/regexp.go#L305) mentions global variables explicitly:
```
// MustCompile is like Compile but panics if the expression cannot be parsed.
// It simplifies safe initialization of global variables holding compiled regular
// expressions.
```
I would be concerned about suggesting code with `regexp.MustCompile` to be rewritten as struct field members. It sounds like it might encourage Singletons structs, and other worse code-readability antipatterns, whereas var members are very readable and easy to find.

---
Obviously that is all just my opinion. If there isn't support for this, then it would be more clear to close this PR, rather than leave it open and confuse others in the future.

**Comment by leighmcculloch** [comment]
Thanks for reviving this PR. I apologize it dropped through the cracks. I think there's a lot of value including regular expression compilation, even though they are not constants, many Go developers treat them like constants including myself.

The errors I'm unclear about, and I think it'd be great if we can make `errX` and `ErrX` only pass if the right-hand side is `errors.New` or `fmt.Errorf` as suggested above too.

Could we reduce this PR with just the `regexp.MustCompile` function in the list?

**Comment by bombsimon** [comment]
Sure thing, I can split this into two separate PRs, probably later today or tomorrow.

File: check_no_globals.go [hunk 3]
```diff
@@ -70,10 +110,27 @@ func checkNoGlobals(rootPath string, includeTests bool) ([]string, error) {
...
+					break
+				}
+
+				if onlyAllowedValues {
+					continue
+				}
+
 				for _, vn := range valueSpec.Names {
-					if isWhitelisted(vn) {
+					if isAllowed(vn) {
```
Comment (leighmcculloch) [inline check_no_globals.go:130]: Thank you for renaming whitelist to allow. 💯

File: main.go [hunk 4]
```diff
@@ -1,4 +1,4 @@
-package main // import "4d63.com/gochecknoglobals"
+package main
```
Comment (leighmcculloch) [inline main.go:1]: This change is unrelated to the PR, can you revert it? Let's discuss issues with this in a new issue if this is causing problems.

```suggestion
package main // import "4d63.com/gochecknoglobals"
```

**Review by leighmcculloch — APPROVED** [review]
🎉 Looks fantastic.

File: testdata/10/code.go [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+	"regexp"
+)
+
+// myVar is just a bad named global var.
+var myVar = 1
+
+// ErrNotFound is an error and should be OK.
+var ErrNotFound = errors.New("this is error")
+
+// ErrIsNotErr is not an error and should not be OK.
```
Comment (bombsimon) [inline testdata/10/code.go:15]:
```suggestion
// ErrIsNotErr is an error and should be OK.
```

Missed this one reverting, this is still ok so the comment is wrong. The tests are passing though.

**Comment by leighmcculloch** [comment]
👏 Thanks @bombsimon!

**Comment by leighmcculloch** [comment]
Closed #6
